### PR TITLE
IISS4 Reward System

### PR DESCRIPTION
**Pull request by Anurupa03**

### CODE REVIEW

File: core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java [hunk 1]
```diff
@@ -100,6 +104,12 @@ public void updatePreps(){
         for (int i = 0; i < totalPreps; i++) {
             this.topPreps.removeLast();
         }
+        List<Address> validPreps = getValidPreps();
+        int size = validPreps.size();
+
```
Comment (goldworm) [inline core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java:109]:
```suggestion

        int size = this.validPreps.size();
```

File: core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java [hunk 2]
```diff
@@ -433,10 +465,49 @@ private void checkForIscore() {
         BigInteger iscoreGenerated = (BigInteger) iscoreDetails.get("estimatedICX");
         if (iscoreGenerated.compareTo(BigInteger.ZERO) > 0) {
             Context.call(SYSTEM_SCORE_ADDRESS, "claimIScore");
+
+            List<Address> topPrep = getTopPreps();
+            List<Address> validPreps = getValidPreps();
+            int size = validPreps.size();
```
Comment (goldworm) [inline core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java:471]:
```suggestion
            int size = this.validPreps.size();
```

**Review by goldworm — CHANGES_REQUESTED** [review]

File: core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java [hunk 3]
```diff
@@ -433,10 +465,49 @@ private void checkForIscore() {
...
+
+        boolean validRate = true;
+        if (commissionRate == null){
+            validRate = true;
+        }  else if (commissionRate.compareTo(getCommissionRate()) > 0) {
+            validRate = false;
+        }
+
+        return notJailed && validRate;
+    }
```
Comment (AntonAndell) [inline core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java:509]: Could avoid double negation here for readability
notJailed to jailed and then use !jailed if negation i needed

File: core-contracts/FeeDistribution/src/main/java/finance/omm/score/fee/distribution/AbstractFeeDistribution.java [hunk 4]
```diff
@@ -74,12 +84,15 @@ protected void distributeFee(BigInteger amount) {
...
 
         Map<String, BigInteger> ommValidators = call(Map.class, Contracts.STAKING,
                 "getActualUserDelegationPercentage", lendingPoolCoreAddr);
         BigInteger amountToDistribute = amount;
         BigInteger denominator = BigInteger.valueOf(100).multiply(ICX);
+
+        BigInteger remaining = BigInteger.ZERO;
+
         for (String validator : ommValidators.keySet()) {
             Address validatorAddr = Address.fromString(validator);
```
[the remark this answers is not in the document]
Comment (Anurupa03) [reply core-contracts/FeeDistribution/src/main/java/finance/omm/score/fee/distribution/AbstractFeeDistribution.java:103]: Got it. I have made the changes [here](https://github.com/ommfinance/openmoneymarket-contracts/commit/fe3b16c37096390362a7fcec9ba42f2dd5e59bb7)

File: core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java (hunk 3 above)
Comment (Anurupa03) [reply core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java]: I have made the changes [here](https://github.com/ommfinance/openmoneymarket-contracts/commit/f0e2dbb60a546e5705270d4b23656cbcb36053df)

File: core-contracts/FeeDistribution/src/main/java/finance/omm/score/fee/distribution/AbstractFeeDistribution.java [hunk 5]
```diff
@@ -88,12 +101,32 @@ private void distributeFeeToValidator(Address lendingPoolCoreAddr,BigInteger amo
```
[the remark this answers is not in the document]
Comment (Anurupa03) [reply core-contracts/FeeDistribution/src/main/java/finance/omm/score/fee/distribution/AbstractFeeDistribution.java:123]: Regarding the calculation of `feePortion`. The equation is : 
BigInteger feePortion = (currentPercentage.multiply(amountToDistribute)).divide(denominator);


- `currentPercentage` represents the delegation percentage by users to validators and is expected to always be greater than 0.001%. [Here](https://github.com/ommfinance/openmoneymarket-contracts/blob/audit/issue-32-fix/core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java#L635)
- The `amountToDistribute` is the value passed from the call to the `distributeFeeToValidator` function. Notably, this function is invoked only when the `amountToDistribute` is greater than zero. [Here](https://github.com/ommfinance/openmoneymarket-contracts/blob/audit/issue-32-fix/core-contracts/FeeDistribution/src/main/java/finance/omm/score/fee/distribution/AbstractFeeDistribution.java#L63)

Given these conditions, it is guaranteed that `feePortion` will always be greater than zero.

**Review by Suyog007 — APPROVED** [review]

**Pull request merged**